### PR TITLE
Symmetry flag and adiabatic BC

### PR DESCRIPTION
**Pull request by shmidt9510**
Introducing three flags. `Adiabatic_BC` could be optional, but `IC_from_temperature` is necessary for me cause there were many cases where I had needed to turn off initial temperature.

### CODE REVIEW

File: slm_porosity/heat_transfer.f90 [hunk 1]
```diff
@@ -230,6 +234,8 @@ SUBROUTINE user_initial_conditions (u, nlocal, ne_local, t_local, scl, scl_fltwt
       !   b) update it by solving the corresponding equation
       func_newton => enthalpy_equation
       u(:,n_var_enthalpy) = find_root(u(:,n_var_enthalpy), temp)
+    ELSEIF ( IC_restart_mode.EQ.0 ) THEN
+      u(:,n_var_enthalpy) = 0.0_pr
```
Comment (olegrog) [inline slm_porosity/heat_transfer.f90]: Do we need this flag? Why we can't use `initial_temperature=0`?

File: slm_porosity/heat_transfer.f90 [hunk 2]
```diff
@@ -256,14 +262,18 @@ SUBROUTINE user_algebraic_BC (Lu, u_in, nlocal, ne_local, jlev, meth)
         IF (ANY( face(1:dim) /= 0) .AND. ie == n_var_enthalpy) THEN
           CALL get_all_indices_by_face(face_type, jlev, nloc, iloc)
           IF (nloc > 0) THEN
-            IF (dim == 3 .AND. face(2) < 0) THEN
-              Lu(shift+iloc(1:nloc)) = du(ie, iloc(1:nloc), 2)
-            ELSEIF (face(dim) > 0) THEN
-              Lu(shift+iloc(1:nloc)) = &
-                Neumann_bc(enthalpy_prev(iloc(1:nloc)), psi_prev(iloc(1:nloc))) * du(ie, iloc(1:nloc), dim) + &
-                Dirichlet_bc(enthalpy_prev(iloc(1:nloc))) * u_in(shift+iloc(1:nloc))
+            IF (.NOT. adiabatic_BC) THEN
```
Comment (olegrog) [inline slm_porosity/heat_transfer.f90]: If you write here `IF (adiabatic_BC) THEN`, your commit will be much shorter. Avoid multilevel `IF ... THEN` if possible. You can add this rule as a guideline.

File: slm_porosity/problem.inp [hunk 3]
```diff
@@ -164,3 +164,6 @@ check_resolution = T                        # if true, abort program in case of
 thickness_points = 8                        # minimum number of points inside the solid--liquid interface thickness
 j_mx_porosity = 7                           # J_MX for porosity
 cfl_fusion_factor = 2                       # the time step will be reduced this number of times
+symmetry_flag = T                           # how the laser heat flux is implemented: T - via BC, F - via RHS
```
Comment (olegrog) [inline slm_porosity/problem.inp]: Do you read your own PR?

File: slm_porosity/heat_transfer.f90 [hunk 4]
```diff
@@ -112,6 +113,9 @@ END FUNCTION func_with_derivative
   INTEGER  :: thickness_points
   INTEGER  :: j_mx_porosity
   REAL(pr) :: cfl_fusion_factor
+  LOGICAL  :: symmetry_flag
```
Comment (olegrog) [inline slm_porosity/heat_transfer.f90]: 1. `flag` is tautology. We already use `LOGICAL` to define this variable as boolean flag.
2. `symmetry` is a quality of the problem, but it is not about, how we deal with this quality.

**Review by olegrog — CHANGES_REQUESTED** [review]
I think we should discuss it.

**Comment by olegrog** [comment]
Where is changes in `user_stats`?

File: slm_porosity/heat_transfer.f90 (hunk 2 above)
Comment (olegrog) [reply slm_porosity/heat_transfer.f90]: Moreover, your style is a potential source of conflicts during rebasing.

File: slm_porosity/heat_transfer.f90 (hunk 1 above)
Comment (shmidt9510) [reply slm_porosity/heat_transfer.f90]: I already told that there are errors with intial_temperature = 0 and even with = 0.02-0.8. If bug would be detected this flag can be deleted.

File: slm_porosity/heat_transfer.f90 (hunk 2 above)
Comment (shmidt9510) [reply slm_porosity/heat_transfer.f90]: I thought about that. It could be added two times more in already existing conditions. But I can't say which way is better.

File: slm_porosity/heat_transfer.f90 [hunk 5]
```diff
@@ -112,6 +113,9 @@ END FUNCTION func_with_derivative
...
+  LOGICAL  :: adiabatic_BC
+  LOGICAL  :: IC_from_temperature
```
Comment (olegrog) [inline slm_porosity/heat_transfer.f90]: Anyway, it a parameter about initial conditions.

File: slm_porosity/heat_transfer.f90 [hunk 6]
```diff
@@ -112,6 +113,10 @@ END FUNCTION func_with_derivative
   INTEGER  :: j_mx_porosity
   REAL(pr) :: cfl_fusion_factor
 
+  ! boundary choosing parameters
```
Comment (olegrog) [inline slm_porosity/heat_transfer.f90]: `boundary conditions`.

File: slm_porosity/heat_transfer.f90 [hunk 7]
```diff
@@ -228,7 +233,6 @@ SUBROUTINE user_initial_conditions (u, nlocal, ne_local, t_local, scl, scl_fltwt
       u(:,n_var_enthalpy) = enthalpy(temp, lf_from_temperature(temp))
       !   b) update it by solving the corresponding equation
       func_newton => enthalpy_equation
-      u(:,n_var_enthalpy) = find_root(u(:,n_var_enthalpy), temp)
```
Comment (olegrog) [inline slm_porosity/heat_transfer.f90]: what?

File: slm_porosity/heat_transfer.f90 [hunk 8]
```diff
@@ -669,8 +681,17 @@ SUBROUTINE user_read_input ()
       PRINT fmt_real, 'conductivity(1) =', conductivity(1.0_pr, 0.5_pr)
       PRINT fmt_real, 'interface thickness =', interface_thickness
     END IF
+
+    CALL check_user_variables
   END SUBROUTINE user_read_input
 
+  SUBROUTINE check_user_variables
```
Comment (olegrog) [inline slm_porosity/heat_transfer.f90]: We intend to check only input parameters, so use `check_user_input`.

File: slm_porosity/heat_transfer.f90 [hunk 9]
```diff
@@ -669,8 +681,17 @@ SUBROUTINE user_read_input ()
...
+    IMPLICIT NONE
+
+    IF ((.NOT.symmetry_half_domain .AND. dim.EQ.3) .AND. &
```
Comment (olegrog) [inline slm_porosity/heat_transfer.f90]: Too long condition => not readable. Please, elaborate carefully.

File: slm_porosity/heat_transfer.f90 [hunk 10]
```diff
@@ -112,6 +113,10 @@ END FUNCTION func_with_derivative
   INTEGER  :: j_mx_porosity
   REAL(pr) :: cfl_fusion_factor
 
+  ! boundary choosing parameters
+  LOGICAL  :: symmetry_half_domain
```
Comment (olegrog) [inline slm_porosity/heat_transfer.f90]: Look too long. Let it be `half_domain` with `symmetry` in comments.

File: slm_porosity/heat_transfer.f90 [hunk 11]
```diff
@@ -669,8 +681,17 @@ SUBROUTINE user_read_input ()
...
+
+    CALL check_user_variables
   END SUBROUTINE user_read_input
 
+  SUBROUTINE check_user_variables
+    IMPLICIT NONE
+
+    IF ((.NOT.symmetry_half_domain .AND. dim.EQ.3) .AND. &
+      (initial_laser_position(2).LE.xyzlimits(1,2) .OR. initial_laser_position(2).GE.xyzlimits(2,2))) &
+      STOP 'Boundary conditions are wrong, check coherense of initial_laser_position and domain limits'
```
Comment (olegrog) [inline slm_porosity/heat_transfer.f90]: Check your grammar first of all.

**Review by olegrog — CHANGES_REQUESTED** [review]
Flag `half_domain` should be taken into account over the whole code. It should not change the results.

File: slm_porosity/problem.inp [hunk 12]
```diff
@@ -151,6 +151,10 @@ initial_temp = 1.8
 initial_pool_radius = 1.8
 initial_laser_position = 0, 0, 0            # starting point of the center of the laser beam
 
+# boundary conditions
+half_domain = T                             # T - symmetry across xz plain at y_max, F - no symmetry
```
Comment (olegrog) [inline slm_porosity/problem.inp]: `# solve a problem that is symmetric with respect to y=0 only for y>0.`

File: slm_porosity/problem.inp [hunk 13]
```diff
@@ -151,6 +151,10 @@ initial_temp = 1.8
...
+adiabatic_BC = F                            # T - adiabatic boundary condition, F - boundary condition for the task
```
Comment (olegrog) [inline slm_porosity/problem.inp]: `# if true, replace all BCs by the adiabatic ones

**Review by olegrog — CHANGES_REQUESTED** [review]

File: slm_porosity/problem.inp [hunk 14]
```diff
@@ -164,5 +169,4 @@ check_resolution = T                        # if true, abort the program if the
 thickness_points = 8                        # minimum number of points inside the solid--liquid interface thickness
 j_mx_porosity = 7                           # J_MX for porosity
 cfl_fusion_factor = 2                       # the time step will be reduced this number of times
-absorption_depth = 0.5
-laser_as_bc = F                             # how the laser heat flux is implemented: T  - via BC, F -  via RHS.
+absorption_depth = 0.5                      # depth of the gaussian distribution
```
Comment (olegrog) [inline slm_porosity/problem.inp]: `depth of the laser heat absorption`
